examples_utils/benchmarks/run_benchmarks.py

```diff
@@ -127,39 +127,50 @@ def run_and_monitor_progress(
     def kill_process(proc_pid: int):
         process = psutil.Process(proc_pid)
         for proc in process.children(recursive=True):
-            logger.info("Killing child process %s" % proc.pid)
+            logger.info("Killing child process %s", proc.pid)
             proc.kill()
-        logger.info("Killing process %s" % proc_pid)
+        logger.info("Killing process %s", proc_pid)
         process.kill()
 
     def proc_thread():
         sel = selectors.DefaultSelector()
         sel.register(proc.stdout, selectors.EVENT_READ)
         sel.register(proc.stderr, selectors.EVENT_READ)
         eof = False
+        decode_error_count = 0
         while not eof:
-            for key, _ in sel.select():
+            selected = 0
+            for key, _ in sel.select(timeout=10):
+                selected += 1
                 stream = key.fileobj
                 data = stream.read1(80)
-                try:
-                    data = data.decode()
-                    if not data:
-                        eof = True
-                    listener.write(data)
-                    listener.flush()
-
-                    if stream is proc.stdout:
-                        outs[0].append(data)
-                    else:
-                        outs[1].append(data)
-                except UnicodeDecodeError as e:
-                    pass
-
-        out, err = proc.communicate()
-        outs[0].append(out.decode())
-        listener.write(out.decode())
-        outs[1].append(err.decode())
-        listener.write(err.decode())
+                data = data.decode(errors="backslashreplace")
+                if not data:
+                    eof = True
+                listener.write(data)
+                listener.flush()
+
+                if stream is proc.stdout:
+                    outs[0].append(data)
+                else:
+                    outs[1].append(data)
+            if not selected:
+                logger.debug("Selector did not pick any files to explore, polling to check for exit")
+                if proc.poll() is not None:
+                    logger.info("Selector did not pick any files to explore, and subprocess has exited. Terminating.")
+                    eof = True
+        try:
+            out, err = proc.communicate(timeout=20)
+            outs[0].append(out.decode())
+            listener.write(out.decode())
+            outs[1].append(err.decode())
+            listener.write(err.decode())
+        except (subprocess.TimeoutExpired, UnicodeDecodeError):
+            proc.poll()
```

Comment:
What is the reason for this line?

Reply:
The reason we call `communicate` is to do three things: collect stdout, stderr and set the exit code. If communicate times out at the end, we still want to make sure the exit code is set, which we can do with `proc.poll()`